File: java/src/tarehart/rlbot/planning/GoalUtil.java
package tarehart.rlbot.planning;

import tarehart.rlbot.Bot;
import tarehart.rlbot.math.BallSlice;
import tarehart.rlbot.math.Plane;
import tarehart.rlbot.math.vector.Vector3;
import tarehart.rlbot.physics.BallPath;

import java.util.Optional;

public class GoalUtil {

    public static final Goal BLUE_GOAL = new Goal(true);
    public static final Goal ORANGE_GOAL = new Goal(false);

    public static Goal getOwnGoal(Bot.Team team) {
        return team == Bot.Team.BLUE ? BLUE_GOAL : ORANGE_GOAL;
    }

    public static Goal getEnemyGoal(Bot.Team team) {
        return team == Bot.Team.BLUE ? ORANGE_GOAL : BLUE_GOAL;
    }

    public static Optional<BallSlice> predictGoalEvent(Goal goal, BallPath ballPath) {
        return ballPath.getPlaneBreak(ballPath.getStartPoint().time, goal.getScorePlane(), true);
    }

    public static boolean ballLingersInBox(Goal goal, BallPath ballPath) {
        Optional<BallSlice> firstSlice = ballPath.findSlice(slice -> goal.isInBox(slice.getSpace()));
        Optional<BallSlice> secondSlice = firstSlice.flatMap(stv -> ballPath.getMotionAt(stv.getTime().plusSeconds(2)));
        return secondSlice.isPresent() && goal.isInBox(secondSlice.get().getSpace());
    }

    public static Goal getNearestGoal(Vector3 position) {
        return position.y > 0 ? ORANGE_GOAL : BLUE_GOAL;
    }
}

File: src/main/java/tarehart/rlbot/planning/GoalUtil.kt
package tarehart.rlbot.planning

import tarehart.rlbot.bots.Team
import tarehart.rlbot.math.Circle
import tarehart.rlbot.math.vector.Vector2
import tarehart.rlbot.math.vector.Vector3
import tarehart.rlbot.physics.ArenaModel
import tarehart.rlbot.physics.BallPath

object GoalUtil {

    private var BLUE_GOAL: Goal = SoccerGoal(true)
    private var ORANGE_GOAL: Goal = SoccerGoal(false)

    fun getOwnGoal(team: Team): Goal {
        return if (team == Team.BLUE) BLUE_GOAL else ORANGE_GOAL
    }

    fun getEnemyGoal(team: Team): Goal {
        return if (team == Team.BLUE) ORANGE_GOAL else BLUE_GOAL
    }

    fun transformNearPost(post: Vector2, ballPosition: Vector2): Vector2 {
        if (ballPosition.x * post.x < 0) {
            // It's actually the far post.
            return post
        }

        val ballCircle = Circle(ballPosition, ArenaModel.BALL_RADIUS.toDouble())

        ballCircle.calculateTangentPoints(post)?.let {
            val riskyTangent = it.toList().sortedBy { it.distance(post) }.first()
            return post + (ballPosition - riskyTangent)
        }

        return post
    }

    fun ballLingersInBox(goal: SoccerGoal, ballPath: BallPath): Boolean {
        val firstSlice = ballPath.findSlice({ slice -> goal.isInBox(slice.space) }, increment = 4)
        val secondSlice = firstSlice?.let { fs -> ballPath.getMotionAt(fs.time.plusSeconds(2.0)) }
        return secondSlice != null && goal.isInBox(secondSlice.space)
    }

    fun getNearestGoal(position: Vector3): Goal {
        return if (position.y > 0) ORANGE_GOAL else BLUE_GOAL
    }

    fun setHoopsGoals() {
        BLUE_GOAL = HoopsGoal(true)
        ORANGE_GOAL = HoopsGoal(false)
    }
}
